### ISSUE TITLE
Preloading an association defined on citext columns can result in a runtime error

### ISSUE THREAD

**Issue by gumchum** [rails/rails#21668]
A runtime error can occur when preloading an association that's defined on two citext columns. It particularly happens when the join key of the two records are case insensitively equal, but case sensitively not equal. This is not taken into account as records are joined in ActiveRecord's association preloader and it results in an mysterious error.

```
class CreateLetters < ActiveRecord::Migration
  def change
    execute "CREATE EXTENSION IF NOT EXISTS citext"

    create_table :as do |t|
      t.citext :name
      t.timestamps null: false
    end

    create_table :cs do |t|
      t.citext :name
      t.timestamps null: false
    end
  end
end

class A < ActiveRecord::Base
  belongs_to :c, foreign_key: :name, primary_key: :name
end

class C < ActiveRecord::Base; end
```

```
2.0.0p353 :001 > A.create(:name => "Mark")
2.0.0p353 :002 > C.create(:name => "mark")
2.0.0p353 :003 > A.preload(:c).to_a
  A Load (0.8ms)  SELECT "as".* FROM "as"
  C Load (1.0ms)  SELECT "cs".* FROM "cs" WHERE "cs"."name" IN ('Mark')
NoMethodError: undefined method `each' for nil:NilClass
        from /home/pair/.rvm/gems/ruby-2.0.0-p353@gateway/gems/activerecord-4.2.4/lib/active_record/associations/preloader/association.rb:93:in `block in associated_records_by_owner'
        from /home/pair/.rvm/gems/ruby-2.0.0-p353@gateway/gems/activerecord-4.2.4/lib/active_record/associations/preloader/association.rb:92:in `each'
        from /home/pair/.rvm/gems/ruby-2.0.0-p353@gateway/gems/activerecord-4.2.4/lib/active_record/associations/preloader/association.rb:92:in `associated_records_by_owner'
        from /home/pair/.rvm/gems/ruby-2.0.0-p353@gateway/gems/activerecord-4.2.4/lib/active_record/associations/preloader/singular_association.rb:9:in `preload'
        from /home/pair/.rvm/gems/ruby-2.0.0-p353@gateway/gems/activerecord-4.2.4/lib/active_record/associations/preloader/association.rb:20:in `run'
        from /home/pair/.rvm/gems/ruby-2.0.0-p353@gateway/gems/activerecord-4.2.4/lib/active_record/associations/preloader.rb:146:in `block (2 levels) in preloaders_for_one'
        from /home/pair/.rvm/gems/ruby-2.0.0-p353@gateway/gems/activerecord-4.2.4/lib/active_record/associations/preloader.rb:144:in `each'
       ....                                                                                                               
```

I can take a shot at fixing this particular part of the code if that seems worthwhile. My current thoughts: if the type of both keys are citext, both keys can be case insensitively compared by downcasing them during the join operation.

**Comment by sgrif** [comment]
Can you please provide a minimal test case using [this template](https://github.com/rails/rails/blob/master/guides/bug_report_templates/active_record_master.rb) that we can use to reproduce the issue?

**Comment by gumchum** [comment]
Here you go. I modified it slightly to use the postgres adapter.

```
 begin
   require 'bundler/inline'
 rescue LoadError => e
   $stderr.puts 'Bundler version 1.10 or later is required. Please update your Bundler'
   raise e
 end
 
 gemfile(true) do
  source 'https://rubygems.org'
  gem 'rails', github: 'rails/rails'
  gem 'arel', github: 'rails/arel'
  gem 'rack', github: 'rack/rack'
  gem 'sprockets', github: 'rails/sprockets'
  gem 'sprockets-rails', github: 'rails/sprockets-rails'
  gem 'sass-rails', github: 'rails/sass-rails'
  gem 'pg'
 end
 
 require 'active_record'
 require 'minitest/autorun'
 require 'logger'
 
 # This connection will do for database-independent bug reports.
 ActiveRecord::Base.establish_connection(adapter: 'postgresql', database: 'test')
 ActiveRecord::Base.logger = Logger.new(STDOUT)
 
 ActiveRecord::Schema.define do
   execute "CREATE EXTENSION IF NOT EXISTS citext"
   create_table :posts, force: true  do |t|
     t.citext :name
   end
 
   create_table :comments, force: true  do |t|
     t.citext :name
   end
 end
 
 class Post < ActiveRecord::Base
   has_many :comments
 end
 
 class Comment < ActiveRecord::Base
   belongs_to :post, foreign_key: :name, primary_key: :name
 end
 
 class BugTest < Minitest::Test
   def test_association_stuff
     post = Post.create!(:name => "hello")
     comment = Comment.create!(:name => "Hello")
 
     begin
       Comment.preload(:post).to_a
     rescue => e
       flunk(e.message)
     end
   end
 end
```

**Comment by sgrif** [comment]
Thanks. Sorry, upon a second read-through I see that your original issue basically already had that. Got caught by my "not using the issue template, probably not enough info to reproduce" autopilot. XD I'll look into it.

**Comment by sgrif** [comment]
Ok, I understand the issue now. This is tricky, and I'm not sure what the right answer is here. As you can probably guess, the reason that the error is occurring is because AR is trying to find the record with the foreign key `"hello"`, when it's actually `"Hello"`. I actually think this is correct, because at this point we're in Ruby-land, and "hello" does not equal "Hello". That also makes me tempted to not fix the error, either, because I think blowing up loudly with "Hey maybe don't do this?" as opposed to not treating those records as associated and potentially silently losing data isn't the right way thing to do. I also definitely don't want to have the association return different results based on whether or not it was preloaded (though you can already make that happen by doing funky things w/ joins...).

TL;DR, while this seems completely reasonable, I'm not sure the association preloader should know or care about the semantics of equality for every type in every adapter, and I'm tempted to just say that this isn't a supported use case.

@matthewd WDYT?

**Comment by gumchum** [comment]
Not fixing and just blowing up with something descriptive seems reasonable to me, especially since it's still possible to include the association with `eager_load`.

**Comment by sgrif** [comment]
:+1: Would you like to write up a patch to improve the error message?

**Comment by matthewd** [comment]
> I'm not sure the association preloader should know or care about the semantics of equality for every type in every adapter

:+1:

**Comment by gaurish** [comment]
This test doesn't fail for me. 

```ruby
 $ ruby error.rb
Updating git://github.com/rails/arel.git
Updating git://github.com/rack/rack.git
Updating git://github.com/rails/sprockets.git
Updating git://github.com/rails/sprockets-rails.git
Updating git://github.com/rails/sass-rails.git
Fetching gem metadata from https://rubygems.org/...........
Fetching version metadata from https://rubygems.org/...
Fetching dependency metadata from https://rubygems.org/..
Resolving dependencies...
Using rake 10.4.2
Using concurrent-ruby 1.0.0.pre5
Using i18n 0.7.0
Using json 1.8.3
Using method_source 0.8.2
Using minitest 5.8.2
Using thread_safe 0.3.5
Using builder 3.2.2
Using erubis 2.7.0
Using mini_portile 0.6.2
Using mime-types 2.6.2
Using arel 7.0.0.alpha from git://github.com/rails/arel.git (at master)
Using bundler 1.10.6
Using pg 0.18.3
Using thor 0.19.1
Using sass 3.4.19
Using rack 2.0.0.alpha from git://github.com/rack/rack.git (at master)
Using tzinfo 1.2.2
Using nokogiri 1.6.6.2
Using mail 2.6.3
Using activesupport 5.0.0.alpha from source at /Users/gaurish/code/repo/rails
Using loofah 2.0.3
Using rails-html-sanitizer 1.0.2
Using rack-test 0.6.3
Using sprockets 4.0.0 from git://github.com/rails/sprockets.git (at master)
Using rails-deprecated_sanitizer 1.0.3
Using globalid 0.3.6
Using activemodel 5.0.0.alpha from source at /Users/gaurish/code/repo/rails
Using rails-dom-testing 1.0.7
Using activejob 5.0.0.alpha from source at /Users/gaurish/code/repo/rails
Using actionview 5.0.0.alpha from source at /Users/gaurish/code/repo/rails
Using activerecord 5.0.0.alpha from source at /Users/gaurish/code/repo/rails
Using actionpack 5.0.0.alpha from source at /Users/gaurish/code/repo/rails
Using actionmailer 5.0.0.alpha from source at /Users/gaurish/code/repo/rails
Using railties 5.0.0.alpha from source at /Users/gaurish/code/repo/rails
Using sprockets-rails 3.0.0.beta3 from git://github.com/rails/sprockets-rails.git (at master)
Using rails 5.0.0.alpha from source at /Users/gaurish/code/repo/rails
Using sass-rails 6.0.0 from git://github.com/rails/sass-rails.git (at master)
-- execute("CREATE EXTENSION IF NOT EXISTS citext")
D, [2015-11-09T21:03:38.270612 #52825] DEBUG -- :    (31.6ms)  CREATE EXTENSION IF NOT EXISTS citext
   -> 0.0636s
-- create_table(:posts, {:force=>true})
D, [2015-11-09T21:03:38.298366 #52825] DEBUG -- :    (7.2ms)  CREATE TABLE "posts" ("id" serial primary key, "name" citext)
   -> 0.0276s
-- create_table(:comments, {:force=>true})
D, [2015-11-09T21:03:38.302779 #52825] DEBUG -- :    (3.3ms)  CREATE TABLE "comments" ("id" serial primary key, "name" citext)
     t.citext :name
   end
 end

 class Post < ActiveRecord::Base
   has_many :comments
 end

 class Comment < ActiveRecord::Base
   belongs_to :post, foreign_key: :name, primary_key: :name
 end

 class BugTest < Minitest::Test
   def test_association_stuff
     post = Post.create!(:name => "hello")
     comment = Comment.create!(:name => "Hello")

     begin
        debugger
       Comment.preload(:post).to_a
     rescue => e
       flunk(e.message)
     end
   end
:0
require 'byebug'
   -> 0.0044s
Run options: --seed 44426

# Running:

D, [2015-11-09T21:03:38.333401 #52825] DEBUG -- :    (0.1ms)  BEGIN
D, [2015-11-09T21:03:38.341820 #52825] DEBUG -- :   SQL (0.7ms)  INSERT INTO "posts" ("name") VALUES ($1) RETURNING "id"  [["name", "hello"]]
D, [2015-11-09T21:03:38.343203 #52825] DEBUG -- :    (0.3ms)  COMMIT
D, [2015-11-09T21:03:38.347384 #52825] DEBUG -- :    (0.1ms)  BEGIN
D, [2015-11-09T21:03:38.348434 #52825] DEBUG -- :   SQL (0.4ms)  INSERT INTO "comments" ("name") VALUES ($1) RETURNING "id"  [["name", "Hello"]]
D, [2015-11-09T21:03:38.348904 #52825] DEBUG -- :    (0.3ms)  COMMIT
D, [2015-11-09T21:03:38.349528 #52825] DEBUG -- :   Comment Load (0.2ms)  SELECT "comments".* FROM "comments"
D, [2015-11-09T21:03:38.358890 #52825] DEBUG -- :   Post Load (0.5ms)  SELECT "posts".* FROM "posts" WHERE "posts"."name" = 'Hello'
.

Finished in 0.035780s, 27.9488 runs/s, 0.0000 assertions/s.

1 runs, 0 assertions, 0 failures, 0 errors, 0 skips
```

**Comment by vipulnsward** [comment]
Changes in https://github.com/rails/rails/commit/6ec7baad52ad0de371194a3696b96ca46b527196 takes care of not raising an error, as well as maintaining that we don't return wrong result.

Here is updated gist, with current behaviour - https://gist.github.com/vipulnsward/fc108d93a1684153bc97105799c49285

Closing, since not an issue not now, and passes, (see above run). Paging @sgrif @matthewd as well.

**Comment by MacksMind** [comment]
This has regressed, and @gumchum's test case now gives undefined method `association' for nil:NilClass

Not sure if anyone cares about a patch to avoid the exception, but I have a workaround to make the preload work. In my case I'm joining email to email on citext database columns. Adding this on both classes fixed the issue for me, and fixes the test case too. (@vipulnsward)

```
def [](attr_name)
  if attr_name == :email &&
     caller(1..1).first.include?('/active_record/associations/preloader/association.rb')
    return read_attribute(:email).downcase
  end

  super
end
```

**Comment by MacksMind** [comment]
Also I'm happy to submit a patch to improve the error...
